Blockchain db pruning horizon limits and orphan pool cleanup

neonknight64 opened the pull request:
## Description
- When a node is loaded, it will check if a different pruning horizon was specified in the node config. If the node was previously running as an archival node then the node can be switched to a pruning node. If it was previously running in pruned mode then the pruning horizon can only be made lower. More advanced mechanisms can later be considered.
- The check_for_valid_heights function was modified to enforce checking of the pruning horizon to allow db queries and operations to be limited when the node is running in pruned mode.
- Orphan cleanup was modified to allow orphan blocks with block heights before the horizon block height to be discarded.
- Cleaning up of the orphan block pool was made more efficient by removing unnecessary hashing of orphan blocks.

## Motivation and Context
- Added a rewind_past_horizon_height test to check the behaviour when the node attempts to rewind the chain past the horizon block height.
- The restore_metadata test was updated to also test configuring of the pruning horizon.
- A horizon_height_orphan_cleanup test was also added to test the cleanup of orphan blocks with lower heights than the horizon block height.

## How Has This Been Tested?
These changes enable pruning horizon limits in the blockchain db and horizon height orphan pool cleanup.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -257,11 +257,21 @@ where T: BlockchainBackend
             blockchain_db.store_new_block(genesis_block)?;
             blockchain_db.store_pruning_horizon(config.pruning_horizon)?;
+        } else if ((metadata.pruning_horizon == 0) && (metadata.pruning_horizon != config.pruning_horizon)) |
+            ((config.pruning_horizon != 0) && (config.pruning_horizon < metadata.pruning_horizon))
```
Should be `||`

sdbondi replied:
Think this can be simplified
```suggestion
        } else if config.pruning_horizon > 0 && (metadata.pruning_horizon == 0 || config.pruning_horizon < metadata.pruning_horizon)
```

"If the config says pruned mode and either the node was in archive mode or the config horizon is less than the current horizon, then update the pruning horizon"

neonknight64 replied:
There was a small logic bug that I fixed. The code should also be much more readable now. The logic should have been: If an archival node then apply configured pruning horizon, or if a pruned node apply configured pruning horizon if less.  The last case might change at a later stage.

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1248,69 +1264,45 @@ fn find_strongest_orphan_tip<T: BlockchainBackend>(
-            txn.delete(DbKey::OrphanBlock(orphans[i].1.clone()));
+        let mut removed_count: usize = 0;
+        for (height, block_hash) in orphans {
+            if (height > horizon_height) & (removed_count >= num_over_limit) {
```
```suggestion
            if height > horizon_height && removed_count >= num_over_limit {
```

Hah... using bitwise `&` on two boolean operands (therefore brackets are needed) and using a boolean `&&` on boolean conditions is the same thing, interesting (though boolean `&&` is more idiomatic)

neonknight64 replied:
Change applied.

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1248,69 +1264,45 @@ fn find_strongest_orphan_tip<T: BlockchainBackend>(
+            txn.delete(DbKey::OrphanBlock(block_hash.clone()));
+            removed_count += 1;
+        }
+        if removed_count > 0 {
```
I would expect commit to not do anything if there are no operations queued up - so code on this level can not have to worry about checking that ?

neonknight64 replied:
Check was removed here and added to the memory_db and lmdb_db backends. The backends were acquiring write locks on zero operation db transactions.

sdbondi reviewed (commented):
Looks good - a few nitty comments

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1248,67 +1264,41 @@ fn find_strongest_orphan_tip<T: BlockchainBackend>(
+            }
+            trace!(target: LOG_TARGET, "Discarding orphan block ({}).", block_hash.to_hex());
+            txn.delete(DbKey::OrphanBlock(block_hash.clone()));
+            removed_count += 1;
```
Not needed anymore

neonknight64 replied:
I think the `removed_count` is still needed to ensure that all the orphan blocks over the storage limit is removed.

Commit:
- Added a rewind_past_horizon_height test to check the behaviour when the node attempts to rewind the chain past the horizon block height.
- The restore_metadata test was updated to also test configuring of the pruning horizon.
- A horizon_height_orphan_cleanup test was also added to test the cleanup of orphan blocks with lower heights than the horizon block height.
- When a node is loaded, it will check if a different pruning horizon was specified in the node config. If the node was previously running as an archival node then the node can be switched to a pruning node. If it was previously running in pruned mode then the pruning horizon can only be made lower. More advanced mechanisms can later be considered.
- The check_for_valid_heights function was modified to enforce checking of the pruning horizon to allow db queries and operations to be limited when the node is running in pruned mode.
- Orphan cleanup was modified to allow orphan blocks with block heights before the horizon block height to be discarded.
- Cleaning up of the orphan block pool was made more efficient by removing unnecessary hashing of orphan blocks.
- These changes enable pruning horizon limits in the blockchain db and horizon height orphan pool cleanup.

sdbondi reviewed (approved):
👍
